File: app/controllers/CarcassonneWebController.java
package controllers;

import de.htwg.cityyanderecarcassonne.Carcassonne;
import de.htwg.cityyanderecarcassonne.controller.ICarcassonneController;
import de.htwg.cityyanderecarcassonne.model.ICard;
import de.htwg.cityyanderecarcassonne.model.IPlayer;
import de.htwg.cityyanderecarcassonne.model.IPosition;
import de.htwg.cityyanderecarcassonne.model.IRegion;
import de.htwg.cityyanderecarcassonne.view.tui.TextUI;
import models.*;
import play.libs.Json;
import play.mvc.Controller;
import play.mvc.Result;
import views.html.cyc;

import java.util.*;

/**
 * This controller passes the commands to the CityYandereCarcassonne game and returns the TUI.
 */
public class CarcassonneWebController extends Controller {
    private final static Carcassonne carcassonne = Carcassonne.getInstance(15, 15, false, true);
    private final static ICarcassonneController controller = carcassonne.getController();

    private Map<IPlayer, String> playerIdMap = new HashMap<>();
    private String lastCardPosition;

    private void execCmd(String cmd) {
        TextUI tui = carcassonne.getTui();
        tui.processInput(cmd);
    }

    public Result cycarcassonne(String cmd) {
        execCmd(cmd);
        return ok(cyc.render(controller));
    }

    public Result cycarcassonneJson(String cmd){
        execCmd(cmd);
        return json();
    }

    public Result json() {
        return ok();
    }

    public Result getGameStatus() {
        return ok(controller.getStatus().toString());
    }

    public Result addPlayer(String name) {
        controller.addPlayer(name);

        IPlayer controllerPlayer = controller.getPlayers().get(controller.getPlayers().size() - 1);

        String pid = "pid_" + playerIdMap.size();
        playerIdMap.put(controllerPlayer, pid);

        Player jsonPlayer = new Player();
        jsonPlayer.pid = pid;
        jsonPlayer.name = name;
        jsonPlayer.meeple = controllerPlayer.getSumMeeples();
        jsonPlayer.score = controllerPlayer.getScore();

        return ok(Json.toJson(jsonPlayer));
    }

    public Result createGame() {
        controller.create();
        controller.startRound();
        return ok();
    }

    public Result getCurrentPlayerId() {
        return ok(playerIdMap.get(controller.getCurrentPlayer()));
    }

    public Result getCurrentCard() {
        return ok(Json.toJson(cardConvert(controller.cardOnHand())));
    }

    public Result rotateCard(String direction) {
        if ("left".equals(direction)) {
            controller.rotateCardLeft();
        } else if ("right".equals(direction)) {
            controller.rotateCardRight();
        } else {
            return badRequest("Unknown direction. Only \"left\" and \"right\" possible.");
        }
        System.out.println(controller.cardOnHand().getLeftMiddle().getID());
        return getCurrentCard();
    }

    private CurrentCard cardConvert(ICard icard) {
        CurrentCard currCard = new CurrentCard();
        currCard.cardname = cardConvertName(icard);
        currCard.orientation = icard.getOrientation();

        return currCard;
    }

    private String cardConvertName(ICard icard) {
        return icard.toString().substring(icard.toString().lastIndexOf(" ") + 1);
    }

    public Result getRemainingCards() {
        return ok(String.valueOf(controller.getRemainingCards()));
    }

    public Result getCardPossibilities() {
        Map<IPosition, String> possMap = controller.getCardPossibilitiesMap(controller.cardOnHand());
        List<PossCardPos> possList = new ArrayList<>();
        for (Map.Entry<IPosition, String> entry : possMap.entrySet()){
            PossCardPos cardPos = new PossCardPos();
            cardPos.selector = entry.getValue();
            cardPos.position = entry.getKey().getY() + "_" + entry.getKey().getX();
            possList.add(cardPos);
        }

        return ok(Json.toJson(possList));
    }

    public Result placeCard(String selection, String position) {
        lastCardPosition = position;
        controller.placeCard(controller.cardOnHand(), selection);
        return getGameStatus();
    }

    public Result getMeeplePossibilities() {
        ICard card = controller.cardOnHand();

        Map<IRegion, String> possMap = controller.getRegionPossibilitiesMap(card);

        TSCard tscard = new TSCard();
        tscard.position = lastCardPosition;
        tscard.name = cardConvertName(card);
        tscard.orientation = card.getOrientation();
        tscard.regions = new HashMap<>();
        for (Map.Entry<IRegion, String> entry : possMap.entrySet()) {
            if (entry.getKey().equals(card.getLeftTop()))
                tscard.regions.put("LT", entry.getValue());
            if (entry.getKey().equals(card.getLeftMiddle()))
                tscard.regions.put("LM", entry.getValue());
            if (entry.getKey().equals(card.getLeftBelow()))
                tscard.regions.put("LB", entry.getValue());

            if (entry.getKey().equals(card.getBelowLeft()))
                tscard.regions.put("BL", entry.getValue());
            if (entry.getKey().equals(card.getBelowMiddle()))
                tscard.regions.put("BM", entry.getValue());
            if (entry.getKey().equals(card.getBelowRight()))
                tscard.regions.put("BR", entry.getValue());

            if (entry.getKey().equals(card.getCenterMiddle()))
                tscard.regions.put("C", entry.getValue());

            if (entry.getKey().equals(card.getTopLeft()))
                tscard.regions.put("TL", entry.getValue());
            if (entry.getKey().equals(card.getTopMiddle()))
                tscard.regions.put("TM", entry.getValue());
            if (entry.getKey().equals(card.getTopRight()))
                tscard.regions.put("TR", entry.getValue());;

            if (entry.getKey().equals(card.getRightTop()))
                tscard.regions.put("RT", entry.getValue());
            if (entry.getKey().equals(card.getRightMiddle()))
                tscard.regions.put("RM", entry.getValue());
            if (entry.getKey().equals(card.getRightBelow()))
                tscard.regions.put("RB", entry.getValue());
        }

        return ok(Json.toJson(tscard));
    }

    public Result placeMeeple(String selection) {
        controller.placeMeeple(controller.getCurrentPlayer(), controller.cardOnHand(), selection);
        return getGameStatus();
    }

    public Result getTownsquare() {
        de.htwg.cityyanderecarcassonne.model.townsquare.Townsquare cts = controller.getTownsquare();

        Townsquare ts = new Townsquare();
        ts.cards = new ArrayList<>();
        for (int x = 0; x < cts.getDimX(); x++) {
            for (int y = 0; y < cts.getDimY(); y++) {
                ICard ctsCard = cts.getCard(x, y);
                if (ctsCard != null) {
                    TSCard card = new TSCard();
                    card.position = y + "_" + x;
                    card.name = cardConvertName(ctsCard);
                    card.orientation = ctsCard.getOrientation();
                    card.regions = new HashMap<>();

                    if (ctsCard.getLeftTop().getPlayer() != null)
                        card.regions.put("LT", playerIdMap.get(ctsCard.getLeftTop().getPlayer()));
                    if (ctsCard.getLeftMiddle().getPlayer() != null)
                        card.regions.put("LM", playerIdMap.get(ctsCard.getLeftMiddle().getPlayer()));
                    if (ctsCard.getLeftBelow().getPlayer() != null)
                        card.regions.put("LB", playerIdMap.get(ctsCard.getLeftBelow().getPlayer()));

                    if (ctsCard.getTopLeft().getPlayer() != null)
                        card.regions.put("TL", playerIdMap.get(ctsCard.getTopLeft().getPlayer()));
                    if (ctsCard.getTopMiddle().getPlayer() != null)
                        card.regions.put("TM", playerIdMap.get(ctsCard.getTopMiddle().getPlayer()));
                    if (ctsCard.getTopRight().getPlayer() != null)
                        card.regions.put("TR", playerIdMap.get(ctsCard.getTopRight().getPlayer()));

                    if (ctsCard.getCenterMiddle().getPlayer() != null)
                        card.regions.put("C", playerIdMap.get(ctsCard.getCenterMiddle().getPlayer()));

                    if (ctsCard.getBelowLeft().getPlayer() != null)
                        card.regions.put("BL", playerIdMap.get(ctsCard.getBelowLeft().getPlayer()));
                    if (ctsCard.getBelowMiddle().getPlayer() != null)
                        card.regions.put("BM", playerIdMap.get(ctsCard.getBelowMiddle().getPlayer()));
                    if (ctsCard.getBelowRight().getPlayer() != null)
                        card.regions.put("BR", playerIdMap.get(ctsCard.getBelowRight().getPlayer()));

                    if (ctsCard.getRightTop().getPlayer() != null)
                        card.regions.put("RT", playerIdMap.get(ctsCard.getRightTop().getPlayer()));
                    if (ctsCard.getRightMiddle().getPlayer() != null)
                        card.regions.put("RM", playerIdMap.get(ctsCard.getRightMiddle().getPlayer()));
                    if (ctsCard.getRightBelow().getPlayer() != null)
                        card.regions.put("RB", playerIdMap.get(ctsCard.getRightBelow().getPlayer()));

                    ts.cards.add(card);
                }
            }
        }

        return ok(Json.toJson(ts));
    }

    public Result finishRound() {
        controller.finishRound();
        return ok();
    }
}

File: app/controllers/CarcassonneWebController.scala
package controllers

import collection.JavaConverters._
import javax.inject.{Inject, Singleton}

import de.htwg.cityyanderecarcassonne.Carcassonne
import de.htwg.cityyanderecarcassonne.controller.ICarcassonneController
import de.htwg.cityyanderecarcassonne.model.ICard
import de.htwg.cityyanderecarcassonne.model.IPlayer
import de.htwg.cityyanderecarcassonne.model.IPosition
import de.htwg.cityyanderecarcassonne.model.IRegion
import de.htwg.cityyanderecarcassonne.view.tui.TextUI
import models._
import play.api.mvc.{AbstractController, ControllerComponents}
import play.api.libs.json._
import play.api.libs.functional.syntax._


import scala.collection.mutable.ListBuffer

@Singleton
class CarcassonneWebController @Inject()(cc: ControllerComponents) extends AbstractController(cc) {
  private val carcassonne: Carcassonne = Carcassonne.getInstance(15, 15, false, true)
  private val controller: ICarcassonneController = carcassonne.getController

  // Json Converters
  implicit val jsonCurrentCardWrites = Json.writes[CurrentCard]
  implicit val jsonPlayerWrites = Json.writes[Player]
  implicit val jsonPossCardPosWrites = Json.writes[PossCardPos]
  implicit val jsonTSCardWrites = Json.writes[TSCard]
  implicit val jsonTownsquareWrites = Json.writes[Townsquare]

  private var playerIdMap: Map[IPlayer, String] = Map()
  private var lastCardPosition: String = _

  private def execCmd(cmd: String): Unit = {
    val tui: TextUI = carcassonne.getTui
    tui.processInput(cmd)
  }

  def cycarcassonne(cmd: String) = Action {
    execCmd(cmd)
    Ok(views.html.cyc.render(controller))
  }

  def getGameStatus = Action {
    okGetGameStatus
  }

  private def okGetGameStatus = Ok(controller.getStatus.toString)

  def addPlayer(name: String) = Action {
    /*implicit val jsonPlayerWrites: Writes[Player] = (
      (JsPath \ "pid").write[String] and
        (JsPath \ "name").write[String] and
        (JsPath \ "meeple").write[String] and
        (JsPath \ "score").write[String]
      )(unlift(Player.unapply))*/
    controller.addPlayer(name)
    val controllerPlayer: IPlayer =
      controller.getPlayers.get(controller.getPlayers.size - 1)
    val pid: String = "pid_" + playerIdMap.size
    playerIdMap += (controllerPlayer -> pid)
    Ok(Json.toJson(convertPlayer(controllerPlayer)))
  }

  def getAllPlayers = Action {
    val players: scala.collection.mutable.Buffer[IPlayer] = controller.getPlayers.asScala
    val jsonPlayers: ListBuffer[Player] = new ListBuffer[Player]
    for (player <- players) {
      jsonPlayers += convertPlayer(player)
    }
    Ok(Json.toJson(jsonPlayers.toList))
  }

  private def convertPlayer(player: IPlayer): Player = {
    val jsonPlayer: Player = Player(
      playerIdMap(player),
      player.getName,
      player.getSumMeeples,
      player.getScore
    )
    jsonPlayer
  }

  def createGame = Action {
    controller.create()
    controller.startRound()
    Ok("ok")
  }

  def getCurrentPlayerId = Action {
    Ok(playerIdMap(controller.getCurrentPlayer))
  }

  def getCurrentCard = Action {
    okGetCurrentCard
  }

  def okGetCurrentCard = Ok(Json.toJson(cardConvert(controller.cardOnHand())))

  def rotateCard(direction: String) = Action {
    if ("left" == direction) {
      controller.rotateCardLeft()
    } else if ("right" == direction) {
      controller.rotateCardRight()
    } else {
      BadRequest("Unknown direction. Only \"left\" and \"right\" possible.")
    }
    println(controller.cardOnHand().getLeftMiddle.getID)
    okGetCurrentCard
  }

  private def cardConvert(icard: ICard): CurrentCard = {
    val currCard: CurrentCard = new CurrentCard(
      cardConvertName(icard),
      icard.getOrientation
    )
    currCard
  }

  private def cardConvertName(icard: ICard): String =
    icard.toString.substring(icard.toString.lastIndexOf(" ") + 1)

  def getRemainingCards() = Action {
    Ok(String.valueOf(controller.getRemainingCards))
  }

  def getCardPossibilities = Action {
    val possMap: scala.collection.mutable.Map[IPosition, String] =
      controller.getCardPossibilitiesMap(controller.cardOnHand()).asScala
    val possList: ListBuffer[PossCardPos] = new ListBuffer()
    for ((key, value) <- possMap) {
      val cardPos: PossCardPos = new PossCardPos(
        value,
        key.getY + "_" + key.getX
      )
      possList += cardPos
    }
    Ok(Json.toJson(possList.toList))
  }

  def placeCard(selection: String, position: String) = Action {
    lastCardPosition = position
    controller.placeCard(controller.cardOnHand(), selection)
    okGetGameStatus
  }

  def getMeeplePossibilities = Action {
    val card: ICard = controller.cardOnHand()

    val possMap: scala.collection.mutable.Map[IRegion, String] = controller.getRegionPossibilitiesMap(card).asScala

    var tmpTsCardMap: Map[String, String] = Map()
    for ((key, value) <- possMap) {
      if (key == card.getLeftTop) tmpTsCardMap += ("LT" -> value)
      if (key == card.getLeftMiddle) tmpTsCardMap += ("LM" -> value)
      if (key == card.getLeftBelow) tmpTsCardMap += ("LB" -> value)

      if (key == card.getBelowLeft) tmpTsCardMap += ("BL" -> value)
      if (key == card.getBelowMiddle) tmpTsCardMap += ("BM" -> value)
      if (key == card.getBelowRight) tmpTsCardMap += ("BR" -> value)

      if (key == card.getCenterMiddle) tmpTsCardMap += ("C" -> value)

      if (key == card.getTopLeft) tmpTsCardMap += ("TL" -> value)
      if (key == card.getTopMiddle) tmpTsCardMap += ("TM" -> value)
      if (key == card.getTopRight) tmpTsCardMap += ("TR" -> value)

      if (key == card.getRightTop) tmpTsCardMap += ("RT" -> value)
      if (key == card.getRightMiddle) tmpTsCardMap += ("RM" -> value)
      if (key == card.getRightBelow) tmpTsCardMap += ("RB" ->  value)
    }

    val tscard: TSCard = new TSCard(
      lastCardPosition,
      cardConvertName(card),
      card.getOrientation,
      tmpTsCardMap
    )

    Ok(Json.toJson(tscard))
  }

  def placeMeeple(selection: String) = Action {
    controller.placeMeeple(controller.getCurrentPlayer, controller.cardOnHand(), selection)
    okGetGameStatus
  }

  def getTownsquare = Action {
    val cts: de.htwg.cityyanderecarcassonne.model.townsquare.Townsquare = controller.getTownsquare

    val tsCardsTmpList: ListBuffer[TSCard] = ListBuffer()
    for (x <- 0 until cts.getDimX; y <- 0 until cts.getDimY) {
      val ctsCard: ICard = cts.getCard(x, y)
      if (ctsCard != null) {
        var tmpTsCardMap: Map[String, String] = Map()
        if (ctsCard.getLeftTop.getPlayer != null)
          tmpTsCardMap += ("LT" -> playerIdMap(ctsCard.getLeftTop.getPlayer))
        if (ctsCard.getLeftMiddle.getPlayer != null)
          tmpTsCardMap += ("LM" -> playerIdMap(ctsCard.getLeftMiddle.getPlayer))
        if (ctsCard.getLeftBelow.getPlayer != null)
          tmpTsCardMap += ("LB" -> playerIdMap(ctsCard.getLeftBelow.getPlayer))

        if (ctsCard.getTopLeft.getPlayer != null)
          tmpTsCardMap += ("TL" -> playerIdMap(ctsCard.getTopLeft.getPlayer))
        if (ctsCard.getTopMiddle.getPlayer != null)
          tmpTsCardMap += ("TM" -> playerIdMap(ctsCard.getTopMiddle.getPlayer))
        if (ctsCard.getTopRight.getPlayer != null)
          tmpTsCardMap += ("TR" -> playerIdMap(ctsCard.getTopRight.getPlayer))

        if (ctsCard.getCenterMiddle.getPlayer != null)
          tmpTsCardMap += ("C" -> playerIdMap(ctsCard.getCenterMiddle.getPlayer))

        if (ctsCard.getBelowLeft.getPlayer != null)
          tmpTsCardMap += ("BL" -> playerIdMap(ctsCard.getBelowLeft.getPlayer))
        if (ctsCard.getBelowMiddle.getPlayer != null)
          tmpTsCardMap += ("BM" ->playerIdMap(ctsCard.getBelowMiddle.getPlayer))
        if (ctsCard.getBelowRight.getPlayer != null)
          tmpTsCardMap += ("BR" -> playerIdMap(ctsCard.getBelowRight.getPlayer))

        if (ctsCard.getRightTop.getPlayer != null)
          tmpTsCardMap += ("RT" -> playerIdMap(ctsCard.getRightTop.getPlayer))
        if (ctsCard.getRightMiddle.getPlayer != null)
          tmpTsCardMap += ("RM" -> playerIdMap(ctsCard.getRightMiddle.getPlayer))
        if (ctsCard.getRightBelow.getPlayer != null)
          tmpTsCardMap += ("RB" -> playerIdMap(ctsCard.getRightBelow.getPlayer))

        val card: TSCard = new TSCard(
          y + "_" + x,
          cardConvertName(ctsCard),
          ctsCard.getOrientation,
          tmpTsCardMap
        )

        tsCardsTmpList += card
      }
    }
    val ts: Townsquare = new Townsquare(tsCardsTmpList.toList)
    Ok(Json.toJson(ts))
  }

  def finishRound()= Action {
    controller.finishRound()
    Ok("ok")
  }
}
